pipeline_steps/common.groovy

```diff
@@ -2131,14 +2127,75 @@ Boolean willOverlapMaintenanceWindow(){
   LocalDateTime maintWindowStart = getNextMaintenanceWindowStart()
   LocalDateTime prBufferStart = maintWindowStart.minus(prBuffer)
 
-  return prBufferStart.isBefore(completion)
+  Boolean willOverlap = prBufferStart.isBefore(completion)
+  if(willOverlap){
+    print("This build is predicted to overlap the next maintenance window")
+  }
+  return willOverlap
 }
 
+
+
 Boolean issueExistsForNextMaintenanceWindow(project="RE"){
   LocalDateTime nextWindow = getNextMaintenanceWindowStart()
   String nextWindowDateString = nextWindow.format(DateTimeFormatter.ISO_LOCAL_DATE)
   List issues = jira_query("project=\"${project}\" AND summary ~ '${nextWindowDateString} Maintenance Window'")
-  return issues.size > 0
+  Boolean issueExists = issues.size > 0
+  if (issueExists){
+    print("Issue exists for next maintenance window: ${issues[0]}")
+  }
+  return issueExists
+}
+
+/**
+* Jobs matching the patterns in this function are allowed to run during
+* a maintenance window to facilitate testing. Note this only overrides
+* the abort check in standard jobs, it won't override quietDown mode.
+*/
+@NonCPS
+Boolean allowBuildDuringMaintenanceWindow(){
+  patterns = [
+    /RE-Maintenance/,
+    /RE-unit-test/,
+    /RPC-Gating-Unit-Tests/,
+    /scratchpipeline/
+  ]
+
+  for (pattern in patterns){
+    if (env.JOB_NAME.contains(pattern)){
+      return true
+    }
+  }
+  print("This Job is not allowed to build during maintenance windows")
+  return false
+}
+
+Boolean maintenanceInProgress(){
+  String query
+  List issues
+  String jiraIssue = ""
+  try {
+    jiraIssue = maintenanceIssueForDate("today")
+  } catch (Exception e){
+    // No maintenance issue for today so no maintenance is in progress.
+    return false
+  }
+
+  Boolean afterStartTime =  LocalTime.now() > LocalTime.of(maintHour, 0)
+  query = "key=\"${jiraIssue}\" and STATUS in (Doing)"
```

Comment:
I think using doing column leaves this open to mistakes, there are potentially reasons that a maintenance card could go into one of the columns between doing and done without the maintenance being over. Using the done column is unambiguous.

What happens if the maintenance window has started but the issue is still in the backlog? Is it intentional that build should be able to run during this time or is there some other mechanism that prevents them from starting?

Reply:
> there are potentially reasons that a maintenance card could go into one of the columns between doing and done without the maintenance being over. Using the done column is unambiguous

You could use ```statusCategory``` rather than a hardcoded ```status```. 
```'statusCategory = "In Progress"'```would  cover any status in between ```Backlog``` and ```Done```.